File: daemon/Sources/ValueGuard/Policy.swift
import Foundation

public enum PolicyAction: UInt8, Sendable {
    case log = 0
    case blur = 1
    case block = 2
}

public struct PolicyCategory: Sendable {
    public let id: String
    public let threshold: Float
    public let action: PolicyAction
    /// L2-normalized "unsafe" caption-ensemble embedding.
    public let positiveEmbedding: [Float]
    /// L2-normalized "safe" caption-ensemble embedding.
    public let negativeEmbedding: [Float]
}

public struct PolicyFlag: Sendable {
    public let category: PolicyCategory
    public let positiveScore: Float
    public let negativeScore: Float
    public let timestamp: Date
}

enum PolicyError: Error {
    case fileTooShort
    case badMagic
    case unsupportedVersion(UInt32)
    case mismatchedDim(expected: Int, got: Int)
    case truncatedCategory
    case invalidAction(UInt8)
    case invalidUTF8
}

public struct Policy: Sendable {
    public let embedDim: Int
    public let categories: [PolicyCategory]

    public init(loadingFrom url: URL) throws {
        let data = try Data(contentsOf: url, options: [.mappedIfSafe])
        guard data.count >= 20 else { throw PolicyError.fileTooShort }

        let (dim, categories) = try data.withUnsafeBytes { raw -> (Int, [PolicyCategory]) in
            let base = raw.baseAddress!.assumingMemoryBound(to: UInt8.self)
            let magic = Data(bytes: base, count: 4)
            guard magic == Data([0x56, 0x47, 0x50, 0x31]) else { throw PolicyError.badMagic }

            // id_utf8 is variable-length, so every field after it can be misaligned.
            // Use loadUnaligned for scalars and memcpy for the float arrays.
            let version = raw.loadUnaligned(fromByteOffset: 4, as: UInt32.self)
            guard version == 1 else { throw PolicyError.unsupportedVersion(version) }

            let nCategories = Int(raw.loadUnaligned(fromByteOffset: 8, as: UInt32.self))
            let dim = Int(raw.loadUnaligned(fromByteOffset: 12, as: UInt32.self))
            // reserved uint32 at offset 16

            var offset = 20
            var categories: [PolicyCategory] = []
            categories.reserveCapacity(nCategories)

            for _ in 0..<nCategories {
                guard offset + 4 <= data.count else { throw PolicyError.truncatedCategory }
                let idLen = Int(raw.loadUnaligned(fromByteOffset: offset, as: UInt32.self))
                offset += 4

                guard offset + idLen <= data.count else { throw PolicyError.truncatedCategory }
                let idData = Data(bytes: base.advanced(by: offset), count: idLen)
                guard let id = String(data: idData, encoding: .utf8) else { throw PolicyError.invalidUTF8 }
                offset += idLen

                guard offset + 8 <= data.count else { throw PolicyError.truncatedCategory }
                let threshold = raw.loadUnaligned(fromByteOffset: offset, as: Float.self)
                offset += 4
                let actionByte = raw.loadUnaligned(fromByteOffset: offset, as: UInt8.self)
                guard let action = PolicyAction(rawValue: actionByte) else {
                    throw PolicyError.invalidAction(actionByte)
                }
                offset += 4 // includes 3 bytes padding

                let vecBytes = dim * MemoryLayout<Float>.size
                guard offset + 2 * vecBytes <= data.count else { throw PolicyError.truncatedCategory }

                var pos = [Float](repeating: 0, count: dim)
                pos.withUnsafeMutableBytes { dst in
                    _ = memcpy(dst.baseAddress!, base.advanced(by: offset), vecBytes)
                }
                offset += vecBytes

                var neg = [Float](repeating: 0, count: dim)
                neg.withUnsafeMutableBytes { dst in
                    _ = memcpy(dst.baseAddress!, base.advanced(by: offset), vecBytes)
                }
                offset += vecBytes

                categories.append(PolicyCategory(
                    id: id,
                    threshold: threshold,
                    action: action,
                    positiveEmbedding: pos,
                    negativeEmbedding: neg
                ))
            }

            return (dim, categories)
        }

        self.embedDim = dim
        self.categories = categories
    }

    /// Score an image embedding against every category and return the categories that fire.
    /// Uses softmax-over-pair: P(unsafe) = exp(pos·x) / (exp(pos·x) + exp(neg·x)).
    public func evaluate(embedding: [Float]) -> [PolicyFlag] {
        precondition(embedding.count == embedDim, "embedding dim mismatch")
        var flags: [PolicyFlag] = []
        let now = Date()
        for cat in categories {
            let pos = dot(cat.positiveEmbedding, embedding)
            let neg = dot(cat.negativeEmbedding, embedding)
            let maxv = max(pos, neg)
            let ep = expf(pos - maxv)
            let en = expf(neg - maxv)
            let pUnsafe = ep / (ep + en)
            if pUnsafe >= cat.threshold {
                flags.append(PolicyFlag(
                    category: cat,
                    positiveScore: pos,
                    negativeScore: neg,
                    timestamp: now
                ))
            }
        }
        return flags
    }
}

@inline(__always)
private func dot(_ a: [Float], _ b: [Float]) -> Float {
    var sum: Float = 0
    let n = a.count
    a.withUnsafeBufferPointer { ap in
        b.withUnsafeBufferPointer { bp in
            for i in 0..<n {
                sum += ap[i] * bp[i]
            }
        }
    }
    return sum
}

File: daemon-py/valueguard_daemon/policy.py
"""VGP1 policy loading and scoring. Port of daemon/Sources/ValueGuard/Policy.swift.

The VGP1 binary format is defined in model-conversion/embed_captions.py and
must be read byte-for-byte identically to the Swift loader (all little-endian):

  magic      : 4 bytes  = b"VGP1"
  version    : uint32   = 1
  n_categories: uint32
  embed_dim  : uint32   = 768
  reserved   : uint32   = 0
  for each category:
    id_len   : uint32
    id_utf8  : id_len bytes
    threshold: float32
    action   : uint8     (0=log, 1=blur, 2=block)
    padding  : 3 bytes
    pos_vec  : embed_dim float32 (L2-normalized)
    neg_vec  : embed_dim float32 (L2-normalized)
"""

from __future__ import annotations

import enum
import struct
from dataclasses import dataclass
from pathlib import Path

import numpy as np

MAGIC = b"VGP1"
SUPPORTED_VERSION = 1
HEADER = struct.Struct("<4sIIII")


class PolicyAction(enum.IntEnum):
    LOG = 0
    BLUR = 1
    BLOCK = 2


class PolicyError(Exception):
    pass


class FileTooShort(PolicyError):
    pass


class BadMagic(PolicyError):
    pass


class UnsupportedVersion(PolicyError):
    pass


class TruncatedCategory(PolicyError):
    pass


class InvalidAction(PolicyError):
    pass


class InvalidUTF8(PolicyError):
    pass


@dataclass(frozen=True)
class PolicyCategory:
    id: str
    threshold: float
    action: PolicyAction
    positive_embedding: np.ndarray  # L2-normalized "unsafe" ensemble, float32
    negative_embedding: np.ndarray  # L2-normalized "safe" ensemble, float32


@dataclass(frozen=True)
class CategoryScore:
    """Per-frame, per-category result, returned for *every* category so
    calibration tooling sees the full score distribution."""

    category: PolicyCategory
    positive_score: float
    negative_score: float
    firing: bool


class Policy:
    def __init__(self, embed_dim: int, categories: list[PolicyCategory]) -> None:
        self.embed_dim = embed_dim
        self.categories = categories

    @classmethod
    def load(cls, path: Path | str) -> "Policy":
        data = Path(path).read_bytes()
        if len(data) < HEADER.size:
            raise FileTooShort(f"{len(data)} bytes")

        magic, version, n_categories, dim, _reserved = HEADER.unpack_from(data, 0)
        if magic != MAGIC:
            raise BadMagic(magic.hex())
        if version != SUPPORTED_VERSION:
            raise UnsupportedVersion(str(version))

        vec_bytes = dim * 4
        offset = HEADER.size
        categories: list[PolicyCategory] = []
        for _ in range(n_categories):
            if offset + 4 > len(data):
                raise TruncatedCategory()
            (id_len,) = struct.unpack_from("<I", data, offset)
            offset += 4

            if offset + id_len > len(data):
                raise TruncatedCategory()
            try:
                cat_id = data[offset : offset + id_len].decode("utf-8")
            except UnicodeDecodeError as exc:
                raise InvalidUTF8() from exc
            offset += id_len

            if offset + 8 > len(data):
                raise TruncatedCategory()
            (threshold,) = struct.unpack_from("<f", data, offset)
            offset += 4
            action_byte = data[offset]
            try:
                action = PolicyAction(action_byte)
            except ValueError as exc:
                raise InvalidAction(str(action_byte)) from exc
            offset += 4  # includes 3 bytes padding

            if offset + 2 * vec_bytes > len(data):
                raise TruncatedCategory()
            pos = np.frombuffer(data, dtype="<f4", count=dim, offset=offset).copy()
            offset += vec_bytes
            neg = np.frombuffer(data, dtype="<f4", count=dim, offset=offset).copy()
            offset += vec_bytes

            categories.append(
                PolicyCategory(
                    id=cat_id,
                    threshold=float(threshold),
                    action=action,
                    positive_embedding=pos,
                    negative_embedding=neg,
                )
            )

        return cls(embed_dim=dim, categories=categories)

    def evaluate(self, embedding: np.ndarray) -> list[CategoryScore]:
        """Score an image embedding against every category.

        Firing semantics: raw cosine, `pos >= threshold`. The negative score
        is computed and logged but does not gate firing — see Policy.swift
        for the calibration history behind dropping the softmax formulation.
        """
        if embedding.shape != (self.embed_dim,):
            raise ValueError(f"embedding dim mismatch: {embedding.shape} != ({self.embed_dim},)")
        scores: list[CategoryScore] = []
        for cat in self.categories:
            pos = float(np.dot(cat.positive_embedding, embedding))
            neg = float(np.dot(cat.negative_embedding, embedding))
            scores.append(
                CategoryScore(
                    category=cat,
                    positive_score=pos,
                    negative_score=neg,
                    firing=pos >= cat.threshold,
                )
            )
        return scores
